iina/PlayerWindowController.swift

```diff
@@ -0,0 +1,516 @@
+    }
+    
+    rightLabel.mode = Preference.bool(for: .showRemainingTime) ? .remaining : .duration
+    
+    updateVolume()
+    
+    observedPrefKeys.forEach { key in
+      UserDefaults.standard.addObserver(self, forKeyPath: key.rawValue, options: .new, context: nil)
+    }
+    
+    notificationCenter(.default, addObserverForName: .iinaFileLoaded, object: player) { [unowned self] _ in
+      self.updateTitle()
+    }
+    
+    NSWorkspace.shared.notificationCenter.addObserver(forName: NSWorkspace.willSleepNotification, object: nil, queue: nil, using: { [unowned self] _ in
+      if Preference.bool(for: .pauseWhenGoesToSleep) {
+        self.player.pause()
+      }
+    })
+  }
+  
+  deinit {
+    ObjcUtils.silenced {
+      for key in self.observedPrefKeys {
+        UserDefaults.standard.removeObserver(self, forKeyPath: key.rawValue)
+      }
+      for (center, observers) in self.notificationObservers {
+        for observer in observers {
+          center.removeObserver(observer)
+        }
+      }
+    }
+  }
+  
+  internal func notificationCenter(_ center: NotificationCenter, addObserverForName name: Notification.Name, object: Any? = nil, using block: @escaping (Notification) -> Void) {
+    let observer = center.addObserver(forName: name, object: object, queue: .main, using: block)
+    notificationObservers[center, default: []].append(observer)
+  }
+
+  internal func setMaterial(_ theme: Preference.Theme) {
+    guard let window = window else { return }
+
+    if #available(macOS 10.14, *) {
+      window.appearance = NSAppearance(iinaTheme: theme)
+    }
+    // See overridden functions for 10.14-
+  }
+  
+  // MARK: - Mouse / Trackpad event
+  
+  override func keyDown(with event: NSEvent) {
+    let keyCode = KeyCodeHelper.mpvKeyCode(from: event)
+    if let kb = PlayerCore.keyBindings[keyCode] {
+      handleKeyBinding(kb)
+    } else {
+      super.keyDown(with: event)
+    }
+  }
+
+  @discardableResult
+  func handleKeyBinding(_ keyBinding: KeyMapping) -> Bool {
+    if keyBinding.isIINACommand {
+      // - IINA command
+      if let iinaCommand = IINACommand(rawValue: keyBinding.rawAction) {
+        handleIINACommand(iinaCommand)
+        return true
+      } else {
+        Logger.log("Unknown iina command \(keyBinding.rawAction)", level: .error)
+        return false
+      }
+    } else {
+      // - mpv command
+      let returnValue: Int32
+      // execute the command
+      switch keyBinding.action[0] {
+      case MPVCommand.abLoop.rawValue:
+        player.abLoop()
+        returnValue = 0
+      default:
+        returnValue = player.mpv.command(rawString: keyBinding.rawAction)
+      }
+      if returnValue == 0 {
+        return true
+      } else {
+        Logger.log("Return value \(returnValue) when executing key command \(keyBinding.rawAction)", level: .error)
+        return false
+      }
+    }
+  }
+  
+  override func mouseUp(with event: NSEvent) {
+    if event.clickCount == 1 {
```

Comment:
Yuck, please use the actual constants if possible: https://developer.apple.com/documentation/appkit/nsevent/eventtype

Reply:
I don't think there's an event for single click so we have to handle it manually.

Reply:
Sorry, I meant to put this on the code that checks `buttonNumber`.

Reply:
Did you mean to use `.otherMouseUp`?
- This one is only for macOS 10.12+, says the documentation (but there's no compile-time check for this)
- It seems like all the buttons on a mouse except the right button and left button are `other mouse`, so it cannot help us to identify whether or not this is a middle button.